### ISSUE TITLE
Add step to verify Corepack installation to avoid "command not found" error

### ISSUE THREAD

**Issue by karlhorky** [upleveled/system-setup#41]
Running `corepack prepare pnpm@latest --activate` can sometimes cause the following error message:

```sh
$ corepack prepare pnpm@latest --activate
zsh: command not found: corepack
```

![Снимок экрана 2023-07-03 в 21 45 57](https://github.com/upleveled/system-setup/assets/1935696/ea07455b-6ed7-452c-9479-825b18f929da)

Let's research why this happens and make changes to avoid it happening:

- [ ] Research when and why this occurs
- [ ] Add a step (or instruction as part of an existing step) to the macOS guide to check for successful Homebrew installation.

**Comment by Boris1605** [comment]
after running those commands in the terminal to upgrade node latest
```bash
brew upgrade node@20
corepack disable
corepack enable
corepack prepare pnpm@latest --activate
```

i have the same problem now that it shows:
```bash
Warning: node@20 20.11.1 already installed
zsh: command not found: corepack
zsh: command not found: corepack
zsh: command not found: corepack
```
<img width="360" alt="Bildschirmfoto 2024-02-15 um 13 17 37" src="https://github.com/upleveled/system-setup/assets/55455069/a37715f0-e38c-4adc-b44c-efd16929dc4e">

**Comment by karlhorky** [comment]
It seems like this can happen after upgrading Node.js via commands like `brew upgrade node@20`.

In that case, we needed to also run:

```bash
brew link --overwrite node@20
```

Added to upgrade cheatsheet:

- https://github.com/upleveled/system-setup/commit/c2e412463d1d18df28969512fb4eb46184b980d2

**Comment by karlhorky** [comment]
@ProchaLu on Windows it appears that this may sometimes be caused on some Windows machines by an execution policy of `Restricted` (the following was on a Windows 10 machine):

![Screenshot 2024-04-29 104040](https://github.com/upleveled/system-setup/assets/1935696/2d28836a-b252-44d8-a513-52349321aec4)

```
PS C: \Windows\system32> Get-ExecutionPolicy - List
        Scope ExecutionPolicy
        ----- ---------------
MachinePolicy Undefined
UserPolicy    Undefined
Process       Undefined
CurrentUser   Undefined
LocalMachine  Undefined

PS C: \Windows\system32> Get-ExecutionPolicy
Restricted

PS C: \Windows\system32> Set-ExecutionPolicy RemoteSigned
Ausführungsrichtlinie ändern
Die Ausführungsrichtlinie trägt zum Schutz vor nicht vertrauenswürdigen Skripts bei. Wenn Sie die Ausführungsrichtlinie
andern,
sind Sie möglicherweise den im Hilfethema
"about_Execution_Policies" unter
"https:/go.microsoft.com/fwlink/?LinkID=135170™ beschriebenen Sicherheitsrisiken ausgesetzt. Möchten Sie die
Ausführungsrichtlinie ändern?
[0] Ja
[A] Ja, alle [N] Nein [K] Nein, keine [H] Anhalten [?] Hilfe (Standard ist "N"): J

PS C:\Windows\system32> Get-ExecutionPolicy
RemoteSigned
```

Before we can change the guide, this needs further research as to:

1. why this execution policy is how it is
2. what best practice is for execution policies
3. and why all other systems have not needed this so far

**Comment by ProchaLu** [comment]
## Windows

[PowerShell's execution policy](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_execution_policies?view=powershell-7.4) is a safety feature that controls the conditions under which PowerShell loads configuration files and runs scripts. This feature helps prevent the execution of malicious scripts. The default execution policy for Windows client computers is `Restricted`. This setting prevents running all script files. 

To allow PowerShell to run scripts, we need to change the setting to `RemoteSigned`, that is the default policy for Windows server computers. This setting allows: 

- Scripts can run
- Requires a digital signature from a trusted publisher on scripts and configuration files that are downloaded from the internet which includes email and instant messaging programs
- Doesn't require digital signatures on scripts that are written on the local computer and not downloaded from the internet

### Best practice 

The best practice is to set the execution policy to `RemoteSigned`. This setting allows you to run scripts that you write on your local computer and scripts that you download from the internet.

### Why didn't all systems have the same execution policy?

For this question I can only speculate right now, but maybe the command `corepack enable` runs a script to [overwrite the existing symlinks](https://nodejs.org/api/corepack.html#:~:text=Enabling%20the%20feature%23&text=To%20do%20that%2C%20run%20corepack,will%20work%20without%20further%20setup.) and that's why the execution policy needs to be changed to `RemoteSigned`?

**Comment by karlhorky** [comment]
we still need some kind of first explanation / investigation for this:

> 3\. and why all other systems have not needed this so far

when a new Windows Home / Pro installation is set up, what is the default? the only two bits of information at [PowerShell's execution policy](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_execution_policies?view=powershell-7.4) linked above are:

> Restricted
>
> - The default execution policy for Windows client computers.

> Default
>
> - Restricted for Windows clients.

are we sure this actually applies to what we're talking about here?

**Comment by ProchaLu** [comment]
When installing Windows for the first time, Windows 10/11 uses PowerShell 5.1. as [default](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_windows_powershell_5.1?view=powershell-5.1#:~:text=Windows%20PowerShell%205.1%20is%20installed%20by%20default%20on%20Windows%20Server%20version%202016%20and%20higher%20and%20Windows%20client%20version%2010%20and%20higher.). The [default execution policy](https://learn.microsoft.com/en-us/powershell/scripting/learn/ps101/01-getting-started?view=powershell-5.1#execution-policy) for Windows 10/11 is `Restricted`, also explained in these different blog posts:
- https://www.hanselman.com/blog/signing-powershell-scripts
- https://www.ninjaone.com/blog/understanding-powershell-execution-policies/
- https://www.mssqltips.com/sqlservertip/2702/setting-the-powershell-execution-policy/

I think this confirms the default behavior. If we don't want to change it, we could change the execution policy for the single `pnpm setup` command because we do the same already when installing Chocolatey. 

```
Set-ExecutionPolicy Bypass -Scope Process -Force;
```

 If a student installs the [latest PowerShell version](https://learn.microsoft.com/en-us/powershell/scripting/learn/ps101/01-getting-started?view=powershell-7.4#execution-policy), the execution policy is also `Restricted`.

**Comment by karlhorky** [comment]
> If we don't want to change it, we could change the execution policy for the single `pnpm setup` command because we do the same already when installing Chocolatey.

this sounds like the best, most secure option, but I think the UX would have a downside here, because it would be required for every time a student installs a new version of pnpm with corepack, right?

**Comment by ProchaLu** [comment]
> this sounds like the best, most secure option, but I think the UX would have a downside here, because it would be required for every time a student installs a new version of pnpm with corepack, right?

Yes, students always have to set the execution policy when running pnpm with corepack. 

Maybe a different solution would be to set the execution policy to `AllSigned`. In the install guide of the [Chocolatey](https://docs.chocolatey.org/en-us/choco/setup/#install-with-powershellexe) docs, we are currently using this install command 
```bash
Set-ExecutionPolicy Bypass -Scope Process -Force; [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072; iex ((New-Object System.Net.WebClient).DownloadString('https://community.chocolatey.org/install.ps1'))
```

This command sets the execution policy to `Bypass` for this single command. As an alternative, the `[Chocolatey docs](https://docs.chocolatey.org/en-us/choco/setup/#install-with-powershellexe)` suggest changing the execution policy to `AllSigned`. `AllSigned` requires that all scripts and configuration files are signed by a trusted publisher, including scripts written on the local computer.

The script to install Chocolatey would be: 
```bash
Set-ExecutionPolicy AllSigned; [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072; iex ((New-Object System.Net.WebClient).DownloadString('https://community.chocolatey.org/install.ps1'))

```

When changing the execution policy in this step, we don't have to make additional changes to the System Setup Guide.
